src/libopensc/card-mcrd.c

```diff
@@ -1391,6 +1395,60 @@ static int mcrd_compute_signature(sc_card_t * card,
 	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, apdu.resplen);
 }
 
+static int mcrd_decipher(struct sc_card *card,
+						 const u8 * crgram, size_t crgram_len,
+						 u8 * out, size_t outlen)
+{
+	sc_security_env_t *env = NULL;
+	int r = 0;
+	size_t sbuf_len = crgram_len + 7;
+	sc_apdu_t apdu;
+	u8 *sbuf = NULL;
+
+	if (card == NULL || crgram == NULL || out == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	env = &DRVDATA(card)->sec_env;
+
+	LOG_FUNC_CALLED(card->ctx);
+	if (env->operation != SC_SEC_OPERATION_DERIVE)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, iso_ops->decipher(card, crgram, crgram_len, out, outlen));
+	if (crgram_len > 255)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, SC_ERROR_INVALID_ARGUMENTS);
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL,
+		 "Will dervie (%d) for %"SC_FORMAT_LEN_SIZE_T"u (0x%02"SC_FORMAT_LEN_SIZE_T"x) bytes using key %d algorithm %d flags %d\n",
+		 env->operation, crgram_len, crgram_len, env->key_ref[0],
+		 env->algorithm, env->algorithm_flags);
+
+	sbuf = malloc(sbuf_len);
+	sbuf[0] = 0xA6; // Control Reference Template Tag for Key Agreement (ISO 7816-4:2013 Table 54)
+	sbuf[1] = 3 + 2 + crgram_len; // Length of the Control reference Template
+	sbuf[2] = 0x7F; // Ephemeral public key Template Tag (ISO 7816-8:2016 Table 3)
+	sbuf[3] = 0x49;
+	sbuf[4] = 2 + crgram_len; // Length of ephemeral public key Template
+	sbuf[5] = 0x86; // External Public Key
+	sbuf[6] = crgram_len; // External Public Key Len
+	memcpy(&sbuf[7], crgram, crgram_len);
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_4_SHORT, 0x2A, 0x80, 0x86);
```

Comment:
above, you're testing for `crgram_len > 255`. Given the encoding, this won't fit in a short length command APDU. Why don't you use `SC_APDU_CASE_4` and let the lower level decide based on the length of data?